File: app/src/main/java/com/dmi/perfectreader/layout/config/TextMetrics.java
package com.dmi.perfectreader.layout.config;

import com.dmi.perfectreader.style.FontStyle;
import com.dmi.util.annotation.Reusable;

public interface TextMetrics {
    @Reusable
    float[] charWidths(CharSequence text, FontStyle style);

    VerticalMetrics verticalMetrics(FontStyle style);

    @Reusable
    class VerticalMetrics {
        protected float ascent;
        protected float descent;
        protected float leading;

        public float ascent() {
            return ascent;
        }

        public float descent() {
            return descent;
        }

        public float leading() {
            return leading;
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/layout/config/TextMetrics.kt
package com.dmi.perfectreader.layout.config

import com.dmi.perfectreader.style.FontStyle
import com.dmi.util.annotation.Reusable

interface TextMetrics {
    @Reusable
    fun charWidths(text: CharSequence, style: FontStyle): FloatArray

    fun verticalMetrics(style: FontStyle): VerticalMetrics

    @Reusable
    open class VerticalMetrics {
        var ascent: Float = 0.toFloat()
        var descent: Float = 0.toFloat()
        var leading: Float = 0.toFloat()

        fun ascent(): Float {
            return ascent
        }

        fun descent(): Float {
            return descent
        }

        fun leading(): Float {
            return leading
        }
    }
}
